src/main/java/net/minecraftforge/common/capabilities/AttachCapabilitiesEvent.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.common.capabilities;
+
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * The AttachCapabilitiesEvent is used to attach {@link IAttachedCapabilityProvider}s to game objects.<br>
+ * The firing of this event is determined by the provider object, and is not consistent across usages.<br>
+ * For most cases, this is fired whenever the {@link CapabilityDispatcher} must be resolved.<br>
+ * This means that typically, this event is not fired until capabilities are attempted to be accessed for a particular object.
+ */
+public abstract class AttachCapabilitiesEvent<T extends ICapabilityProvider> extends Event
+{
+
+    protected final T obj;
+    protected final Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> providers = new HashMap<>();
```

Comment:
I would prefer a multimap (map to list is fine too), it is nicer for dynamic caps, and still yields high performance improvements.

Reply:
Without a reasonable use-case I'm hesitant to allow multiple providers control over the same capability type, since you can induce cases where a provider can attach but silently fail due to another earlier provider providing that capability type.

Reply:
This is useful in case multiple dynamic providers are attached.

Reply:
Well that's my point, I don't think we should allow multiple providers for a single type.

Reply:
Well once again my GenericInternalInventory to IItemHandler adapter would like to disagree. :P